src/webots/control/WbControlledWorld.cpp

```diff
@@ -219,9 +219,22 @@ void WbControlledWorld::addControllerConnection() {
       robotName = buffer;
       delete[] buffer;
     }
+    int nbExternalController = 0;
     foreach (WbRobot *const robot, robots()) {
-      if (robot->isControllerStarted())
-        continue;
+      if (robot->controllerName() == "<extern>")
+	    nbExternalController++;
+    }
+    foreach (WbRobot *const robot, robots()) {
+      if (robot->isControllerStarted()) {
+        // if a specific robot have been targeted by giving a name, or only one external controller is in the scene (hence it's the target)
+        if (robot->controllerName() == "<extern>" && (robotName == robot->name() || nbExternalController == 1)) {
```

Comment:
This test is wrong because if there is only one controller but the robotName doesn't match, the controller should not be disconnected.
This is important as we want to have the exact same behavior as at first connect.
```suggestion
        if (robot->controllerName() == "<extern>" && (robotName == robot->name() || (robotName.isEmpty() && nbExternalController == 1))) {
```